### PR TITLE
THRIFT-5757 Unit tests for php lib

### PR DESCRIPTION
**Pull request by sveneld**
Next step is writing test for php lib.

It will be several pull request for this task.
  
- [X] Did you create an [Apache Jira](https://issues.apache.org/jira/projects/THRIFT/issues/) ticket? [ ([Request account here](https://selfserve.apache.org/jira-account.html), not required for trivial changes)](https://issues.apache.org/jira/browse/THRIFT-5757)
- [X] If a ticket exists: Does your pull request title follow the pattern "THRIFT-NNNN: describe my issue"?
- [X] Did you squash your changes to a single commit?  (not required, but preferred)
- [X] Did you do your best to avoid breaking changes?  If one was needed, did you label the Jira ticket with "Breaking-Change"?
- [ ] If your change does not involve any code, include `[skip ci]` anywhere in the commit message to free up build resources.

### CODE REVIEW

**Comment by sveneld** [comment]
@pkvach it's a beginning of code coverage for folder Protocol. It's much code there, so I will make pull requests step by step

File: .github/workflows/build.yml [hunk 1]
```diff
@@ -120,11 +120,11 @@ jobs:
...
-          compiler/cpp/thrift --gen php:json -r --out ./lib/php/test/Resources/packages/phpjs lib/php/test/Resources/ThriftTest.thrift
-          compiler/cpp/thrift --gen php:classmap,server,rest -r --out ./lib/php/test/Resources/packages/phpcm lib/php/test/Resources/ThriftTest.thrift
+          compiler/cpp/thrift --gen php:nsglobal="Basic" -r --out ./lib/php/test/Resources/packages/php lib/php/test/Resources/ThriftTest.thrift
+          compiler/cpp/thrift --gen php:validate,nsglobal="Validate" -r --out ./lib/php/test/Resources/packages/phpv lib/php/test/Resources/ThriftTest.thrift
+          compiler/cpp/thrift --gen php:validate,oop,nsglobal="ValidateOop" -r --out ./lib/php/test/Resources/packages/phpvo lib/php/test/Resources/ThriftTest.thrift
+          compiler/cpp/thrift --gen php:json,nsglobal="Json" -r --out ./lib/php/test/Resources/packages/phpjs lib/php/test/Resources/ThriftTest.thrift
+          compiler/cpp/thrift --gen php:classmap,server,rest,nsglobal="Classmap" -r --out ./lib/php/test/Resources/packages/phpcm lib/php/test/Resources/ThriftTest.thrift
```
Comment (sveneld) [inline .github/workflows/build.yml:127]: In php class can not be reloaded, so i add global namespace for each test package

File: lib/php/lib/Protocol/TJSONProtocol.php [hunk 2]
```diff
@@ -197,68 +195,6 @@ public function readJSONSyntaxChar($b)
...
-
-    private function hexChar($val)
-    {
-        return dechex($val);
-    }
-
-    private function hasJSONUnescapedUnicode()
-    {
-        if (PHP_MAJOR_VERSION > 5 || (PHP_MAJOR_VERSION == 5 && PHP_MINOR_VERSION >= 4)) {
-            return true;
```
Comment (sveneld) [inline lib/php/lib/Protocol/TJSONProtocol.php:221]: minimal version of php now is 7 so this code will never be used

File: lib/php/lib/Protocol/TJSONProtocol.php [hunk 3]
```diff
@@ -297,21 +233,14 @@ private function writeJSONDouble($num)
             $this->trans_->write(self::QUOTE);
         }
 
+        #TODO add compatibility with for NAN and INF
```
Comment (sveneld) [inline lib/php/lib/Protocol/TJSONProtocol.php:236]: if we will try to make json_encode on such values we will get a silent error, whic can be readed only by method json_get_last_error. In future should be added flag THROW_ERROR_ON_EXCEPTION

File: lib/php/test/Integration/Lib/ClassLoader/ThriftClassLoaderTest.php [hunk 4]
```diff
@@ -0,0 +1,139 @@
...
+
+use phpmock\phpunit\PHPMock;
+use PHPUnit\Framework\TestCase;
+use Thrift\ClassLoader\ThriftClassLoader;
+
+/***
+ * This test depends on running the compiler against the ./Resources/ThriftTest.thrift file:
+ * lib/php/test$ ../../../compiler/cpp/thrift --gen php:classmap,server,rest,nsglobal="Classmap" -r  --out ./Resources/packages/phpcm ./Resources/ThriftTest.thrift
+ */
+class ThriftClassLoaderTest extends TestCase
```
Comment (sveneld) [inline lib/php/test/Integration/Lib/ClassLoader/ThriftClassLoaderTest.php:32]: Separate ThriftClassLoaderTest on Unit and Integration test

File: lib/php/test/Integration/Lib/Protocol/TJSONProtocolTest.php [hunk 5]
```diff
@@ -0,0 +1,638 @@
...
+    {
+        $this->transport = new TMemoryBuffer();
+        $this->protocol = new TJSONProtocol($this->transport);
+        $this->transport->open();
+    }
+
+    /**
+     * @dataProvider writeDataProvider
+     */
+    public function testWrite(
```
Comment (sveneld) [inline lib/php/test/Integration/Lib/Protocol/TJSONProtocolTest.php:62]: change original class from using separate class of Fixtures to using data provider. It's more readable

**Comment by sveneld** [comment]
I add folder test\Integration for tests which depends on code which is generated by thrift library

File: lib/php/lib/Protocol/TJSONProtocol.php (hunk 3 above)
Comment (sveneld) [reply lib/php/lib/Protocol/TJSONProtocol.php:236]: https://issues.apache.org/jira/browse/THRIFT-5770

**Comment by sveneld** [comment]
Write unit tests for protocol is much complicated %)
Currently  code coverage is 68% image
Next step will be restoring of test between different languages.

File: lib/php/test/Integration/Lib/Serializer/JsonSerializeTest.php [hunk 6]
```diff
@@ -27,21 +27,13 @@
```
Comment (pkvach) [inline lib/php/test/Integration/Lib/Serializer/JsonSerializeTest.php:27]: Looks like we need to change the namespace here:
`Test\Thrift\Unit`
->
`Test\Thrift\Integration\Lib\Serializer`

Comment (sveneld) [reply lib/php/test/Integration/Lib/Serializer/JsonSerializeTest.php:27]: ++

File: lib/php/test/Integration/Lib/ClassLoader/ThriftClassLoaderTest.php [hunk 7]
```diff
@@ -0,0 +1,139 @@
...
+     */
+    public function testRegisterDefinition(
+        $definitions,
+        $class,
+        $checkInterfaceExist = false,
+        $useApcu = false,
+        $apcuPrefix = null
+    ) {
+        $this->getFunctionMock('Thrift\ClassLoader', 'apcu_fetch')
+             ->expects($useApcu ? $this->any() : $this->any())
```
[the remark this answers is not in the document]
Comment (sveneld) [reply lib/php/test/Integration/Lib/ClassLoader/ThriftClassLoaderTest.php]: ++

File: lib/php/lib/Protocol/TSimpleJSONProtocol.php [hunk 8]
```diff
@@ -115,6 +115,7 @@ private function writeJSONDouble($num)
             $this->trans_->write(self::QUOTE);
         }
 
+        #TODO add compatibility with for NAN and INF
```
[the remark this answers is not in the document]
Comment (sveneld) [reply lib/php/lib/Protocol/TSimpleJSONProtocol.php]: ++

File: lib/php/test/Unit/Lib/Protocol/TMultiplexedProtocolTest.php [hunk 9]
```diff
@@ -0,0 +1,85 @@
...
+        $expectedName
+    ) {
+        $protocol = $this->createMock(TProtocol::class);
+        $multiplexerPRotocol = new TMultiplexedProtocol($protocol, $serviceName);
+
+        $protocol->expects($this->once())
+            ->method('writeMessageBegin')
+            ->with($expectedName, $type, $seqid);
+
+        $multiplexerPRotocol->writeMessageBegin($name, $type, $seqid);
```
[the remark this answers is not in the document]
Comment (sveneld) [reply lib/php/test/Unit/Lib/Protocol/TMultiplexedProtocolTest.php]: ++

**Comment by pkvach** [comment]
LGTM :+1:

**Comment by sveneld** [comment]
@Jens-G let's merge

**Pull request closed without merging**